### PR TITLE
many: bind and check keys against models when using FDE hooks v2

### PR DESCRIPTION
**Pull request by pedronis**
See last 4 commits.

This is built on top of reworked #10149.

### CODE REVIEW

**Review by mvo5 — APPROVED** [review]
Thank you

File: tests/lib/fde-setup-hook/fde-setup.go [hunk 1]
```diff
@@ -130,23 +155,37 @@ func runFdeRevealKey() error {
 		return fmt.Errorf("fde-reveal-key input is not valid base64: %v", err)
```
Comment (stolowski) [inline tests/lib/fde-setup-hook/fde-setup.go:155]: "fde-reveal-key **key** input is not valida base64: %v" to follow the other error messages?

File: tests/lib/fde-setup-hook/fde-setup.go [hunk 2]
```diff
@@ -130,23 +155,37 @@ func runFdeRevealKey() error {
...
+		// implementing a real world, production grade FDE
+		// hook, the lock operation must be implemented here
+		// to block decryption operations. This example does
+		// nothing.
 	case "features":
-		// XXX: Not used right now but might in the future?
-		fmt.Fprintf(os.Stdout, `{"features":[]}`)
+		fmt.Fprintf(osStdout, `{"features":[]}`)
 	default:
 		return fmt.Errorf(`unsupported operations %q`, js.Op)
```
Comment (stolowski) [inline tests/lib/fde-setup-hook/fde-setup.go:190]: Is plural "operations" intended?

File: tests/lib/fde-setup-hook/fde-setup.go [hunk 3]
```diff
@@ -157,6 +196,9 @@ func runFdeRevealKey() error {
 func main() {
 	var err error
 
+	// XXX: workaround systemd bug
```
Comment (stolowski) [inline tests/lib/fde-setup-hook/fde-setup.go:199]: What bug is it, can you link to a bug# or elaborate?

File: secboot/secboot_sb_test.go [hunk 4]
```diff
@@ -1175,9 +1175,22 @@ func (s *secbootSuite) TestUnlockVolumeUsingSealedKeyIfEncryptedFdeRevealKeyErr(
 	defaultDevice := "name"
 	mockSealedKeyFile := makeMockSealedKeyFile(c, nil)
 
+	restore = secboot.MockSbActivateVolumeWithKeyData(func(volumeName, sourceDevicePath string, keyData *sb.KeyData, options *sb.ActivateVolumeOptions) (sb.SnapModelChecker, error) {
+		// XXX: this is what the real
+		// MockSbActivateVolumeWithKeyData will do
+		// XXX aux key
```
[the remark this answers is not in the document]
Comment (pedronis) [reply secboot/secboot_sb_test.go]: I removed it in a subsequent commit, when I actually implemented generating/setting one, I don't think there are comments like this anymore in the final PR state?

[the remark this answers is not in the document]
Comment (mvo5) [reply secboot/secboot_sb_test.go]: Uh, sorry, I was going over the individual commits, my bad

File: tests/lib/fde-setup-hook/fde-setup.go (hunk 3 above)
Comment (mvo5) [reply tests/lib/fde-setup-hook/fde-setup.go:199]: I can probably remove it, let me try - the bug is reported as https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1921145 but it's only an issue when the hook fails.

Comment (pedronis) [reply tests/lib/fde-setup-hook/fde-setup.go:199]: well, if people use this as an example they, will hit this, so better to leave it

Comment (stolowski) [reply tests/lib/fde-setup-hook/fde-setup.go:199]: Would be good to enhance the comment to say what bug we're working around here (if we leave XXX).

File: tests/lib/fde-setup-hook/fde-setup.go (hunk 2 above)
Comment (mvo5) [reply tests/lib/fde-setup-hook/fde-setup.go:190]: Nice catch, I don't think this was intentional

File: kernel/fde/reveal_key.go [hunk 5]
```diff
@@ -40,8 +40,10 @@ import (
 type RevealKeyRequest struct {
 	Op string `json:"op"`
 
-	SealedKey []byte `json:"sealed-key,omitempty"`
-	KeyName   string `json:"key-name,omitempty"`
+	SealedKey []byte           `json:"sealed-key,omitempty"`
+	Handle    *json.RawMessage `json:"handle,omitempty"`
+	// depracated for v1
```
Comment (stolowski) [inline kernel/fde/reveal_key.go:45]: Typo,
```suggestion
	// deprecated for v1
```

File: tests/nested/manual/uc20-fde-hooks-v1/task.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+summary: Check that the fde-setup hooks work
```
Comment (stolowski) [inline tests/nested/manual/uc20-fde-hooks-v1/task.yaml:1]: The PR is about hooks v2, but the test is about hooks v1, is there a v2 test coming separately?

**Review by stolowski — APPROVED** [review]
LGTM with a caveat that I cannot see the "big picture" for these changes.

File: kernel/fde/reveal_key.go (hunk 5 above)
Comment (mvo5) [reply kernel/fde/reveal_key.go:45]: Thanks, I think we fix this in the followup

File: tests/nested/manual/uc20-fde-hooks-v1/task.yaml (hunk 6 above)
Comment (mvo5) [reply tests/nested/manual/uc20-fde-hooks-v1/task.yaml:1]: This should also be clarified in the followup, the "uc20-fde-hooks" test is the v2 test and v2 is really what everyone should be using from this PR on. I will update the test description to make this clearer

**Comment by mvo5** [comment]
Errors are unrelated

**Pull request merged**